packages/api/src/webhooks/cannon.ts

```diff
@@ -290,43 +375,102 @@ export default class WebhookCannon {
       if (webhook.sharedSecret) {
         let signature = sign(params.body, webhook.sharedSecret);
-        params.headers["Livepeer-Signature"] = `t=${timestamp},v1=${signature}`;
+        params.headers[SIGNATURE_HEADER] = `t=${timestamp},v1=${signature}`;
+      }
+      try {
+        await this.storeTriggerTime(webhook);
+      } catch (e) {
+        console.log(
+          `Unable to store trigger time of webhook ${webhook.id} url: ${webhook.url}`
+        );
       }
 
+      let resp: Response;
+      let errorMessage;
+      let statusCode;
+
       try {
         logger.info(`webhook ${webhook.id} firing`);
         const startTime = process.hrtime();
-        let resp = await fetchWithTimeout(webhook.url, params);
-        await this.storeResponse(webhook, event, resp, startTime);
+        resp = await fetchWithTimeout(webhook.url, params);
+        const responseBody = await resp.text();
+        await this.storeResponse(webhook, event, resp, startTime, responseBody);
+        statusCode = resp.status;
+
         if (resp.status >= 200 && resp.status < 300) {
           // 2xx requests are cool. all is good
           logger.info(`webhook ${webhook.id} fired successfully`);
           return true;
         }
-
+        errorMessage = responseBody;
         if (resp.status >= 500) {
-          await this.retry(trigger);
+          await this.retry(
+            trigger,
+            params,
+            new Error("Status code: " + resp.status)
+          );
         }
 
         console.error(
           `webhook ${webhook.id} didn't get 200 back! response status: ${resp.status}`
         );
         // we don't retry on non 400 responses. only on timeouts
         // this.retry(event);
-        return;
       } catch (e) {
         console.log("firing error", e);
-        await this.retry(trigger);
+        errorMessage = e.message;
+        await this.retry(trigger, params, e);
+      } finally {
+        if (statusCode >= 300 || !statusCode) {
+          try {
+            await this.storeWebhookFailure(
+              trigger.webhook,
+              statusCode,
+              errorMessage
+            );
+          } catch (e) {
+            console.log(
+              `Unable to store failure of webhook ${webhook.id} url: ${webhook.url}`
+            );
+          }
+        }
         return;
       }
     }
   }
 
+  async storeTriggerTime(webhook: DBWebhook) {
+    await this.db.webhook.update(webhook.id, {
+      status: {
+        lastTriggerTime: Date.now(),
+        lastFailure: webhook.status?.lastFailure,
+      },
+    });
+  }
+
+  async storeWebhookFailure(
+    webhook: DBWebhook,
+    statusCode?: number,
+    errorMessage?: string
+  ) {
+    await this.db.webhook.update(webhook.id, {
+      status: {
+        lastTriggerTime: webhook.status?.lastTriggerTime,
+        lastFailure: {
+          timestamp: Date.now(),
+          statusCode: statusCode,
+          errorMessage: errorMessage,
+        },
+      },
```

Comment:
Same, either remove the `lastTriggerTime` or use a spread (`...webhook.status`) operator instead.

If you do change to declare the whole `lastFailure` object on the main logic above, this might also get cleaner cause you only need to set the whole `lastFailure` obj 😁